pipeline/review/main.go

```diff
@@ -76,60 +55,97 @@ func main() {
+
+	// Initialize beam. This must be done after flag.Parse(), but before any flag
+	// validation.
 	// https://cloud.google.com/dataflow/docs/guides/setting-pipeline-options#CreatePipelineFromArgs
 	beam.Init()
-	// parse commandline arguments into config
-	pipelineConfig, err := getConfigFromFlags(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to construct pipeline config: %w", err)
+
+	// Validate inputs.
+	var merr error
+	if *flagGitHubToken == "" && *flagGitHubAppID == "" {
+		merr = errors.Join(merr, fmt.Errorf("one of github token or github app id are required"))
 	}
-	// construct and execute the pipeline
-	pipeline := review.NewCommitApprovalPipeline(pipelineConfig)
-	if err := beamx.Run(ctx, pipeline); err != nil {
-		return fmt.Errorf("failed to execute pipeline: %w", err)
+	if *flagGitHubToken != "" && *flagGitHubAppID != "" {
+		merr = errors.Join(merr, fmt.Errorf("only one of github token or github app id are allowed"))
 	}
-	return nil
-}
 
-// getConfigFromFlags returns review.CommitApprovalPipelineConfig struct
-// using flag values. returns an error if any of the flags have malformed
-// data.
-func getConfigFromFlags(ctx context.Context) (*review.CommitApprovalPipelineConfig, error) {
-	githubTokenSupplier, err := getGitHubTokenSupplier(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to construct token supplier: %w", err)
-	}
-	token, err := githubTokenSupplier.GitHubToken(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get GitHub token: %w", err)
+	// Create qualified table names.
+	pushEventsTable, err := newQualifiedTableName(*flagPushEventsTable)
+	merr = errors.Join(merr, err)
+	commitReviewStatusTable, err := newQualifiedTableName(*flagCommitReviewStatusTable)
+	merr = errors.Join(merr, err)
+	issuesTable, err := newQualifiedTableName(*flagIssuesTable)
+	merr = errors.Join(merr, err)
+
+	// Create the GitHub token source.
+	var githubTokenSource githubauth.TokenSource
+	if *flagGitHubToken != "" {
+		githubTokenSource, err = githubauth.NewStaticTokenSource(*flagGitHubToken)
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("failed to create github static token source: %w", err))
+		}
+	} else {
+		sm, err := secretmanager.NewClient(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to create secret manager client: %w", err)
+		}
+		defer sm.Close()
+
+		privateKeyPEM, err := secrets.AccessSecret(ctx, sm, *flagGitHubAppPrivateKeyResourceName)
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("failed to get github private key pem: %w", err))
+		}
+
+		githubTokenSource, err = githubauth.NewAppTokenSource(*flagGitHubAppID, *flagGitHubAppInstallationID, privateKeyPEM, githubauth.ForAllRepos())
+		if err != nil {
+			merr = errors.Join(merr, fmt.Errorf("failed to create github app token source: %w", err))
+		}
 	}
-	qualifiedPushEventsTable, err := newQualifiedTableName(pushEventsTable)
+	githubToken, err := githubTokenSource.GitHubToken(ctx)
```

Comment:
If we failed to get the githubTokenSource, we should probably return with the list of errors rather than try and use the value. Adding more errors to the multierror will only muddy up the real error in the logs.

Reply:
Also  you will lose your error context as this will be a nill poitner derefrence if my brief glance at the source is correct.

There are likely other similar issues here related to using values that previously generated failures.

I think it might make sense to check for merr being set after processing all the flags and return failure at that point before going forward.

Reply:
Good point, I wrapped the error and return immediately instead.